beacon-chain/types/crystallized_state.go

```diff
@@ -241,33 +280,57 @@ func (c *CrystallizedState) CalculateNewCrystallizedState(aState *ActiveState, s
 
 	// Construct new crystallized state for cycle transition.
 	newCrystallizedState := NewCrystallizedState(&pb.CrystallizedState{
+		ShardAndCommitteesForSlots: c.data.ShardAndCommitteesForSlots,
+		CrosslinkingStartShard:     c.data.CrosslinkingStartShard,
+		DynastyStart:               c.data.DynastyStart,
+		CurrentDynasty:             c.data.CurrentDynasty,
+		DynastySeed:                c.data.DynastySeed,
 		Validators:                 rewardedValidators,
 		LastStateRecalc:            lastStateRecalc + params.CycleLength,
-		ShardAndCommitteesForSlots: newShardAndCommitteesForSlots,
 		LastJustifiedSlot:          justifiedSlot,
 		JustifiedStreak:            justifiedStreak,
 		LastFinalizedSlot:          finalizedSlot,
-		CrosslinkingStartShard:     0, // TODO(#494): Stub. Need to see where this epoch left off.
 		CrosslinkRecords:           newCrossLinkRecords,
-		DynastySeedLastReset:       c.data.DynastySeedLastReset, // TODO(#267): Stub. Need to implement dynasty transition.
 		TotalDeposits:              nextCycleBalance,
 	})
 
 	return newCrystallizedState, nil
 }
 
-func (c *CrystallizedState) calculateNewShardAndCommitteesForSlots() ([]*pb.ShardAndCommitteeArray, error) {
-	newCommittees, err := casper.ShuffleValidatorsToCommittees(
-		c.DynastySeed(),
-		c.Validators(),
-		c.CurrentDynasty(),
-		c.CrosslinkingStartShard(),
+// NewDynastyRecalculations recomputes the validator set. This method is called during a dynasty transition.
+func (c *CrystallizedState) NewDynastyRecalculations(seed [32]byte) (*CrystallizedState, error) {
```

Comment:
Why not just include this as part of `NewStateRecalculations`?

Reply:
We could do that, but for clarity, I'd like those 2 separated. 
State re-calculation is for - applying validators rewards/penalties , process crosslinks, try to establish a finalized slot
Dynasty re-calculation is for - induct new validators, exit old validators and shuffle validators based on randao seed

To enter state transition we need:
`slot - last_state_recalc >= CYCLE_LENGTH`

To enter dynasty transition we need:
```
block.slot_number - crystallized_state.dynasty_start >= MIN_DYNASTY_LENGTH
last_finalized_slot > dynasty_start
For every shard S in shard_and_committee_for_slots, crosslink_records[S].slot > dynasty_start
```

I'm worry there could be cases where we could get in to dynasty transition but not state transition

Reply:
Its not possible with the current spec since `last_finalized_slot` only advances during a state transition. Its a minor point though, we can always move this around later.